File: server/dotnet/sdk-dotnet-example/Business/Services/PaymentService.cs
using System.Net;
using Business.DTOs.AdditionalPaymentActions;
using Business.DTOs.CreatePayments;
using Business.DTOs.GetPaymentDetails;
using Business.Exceptions;
using Business.Interfaces.Handlers;
using Business.Interfaces.SDKClients;
using Business.Interfaces.Services;

namespace Business.Services;

public class PaymentService(IEnumerable<IPaymentMethodHandler> handlers, IPaymentClient paymentClient)
    : IPaymentService
{
    public async Task<CreatePaymentResponseDto> CreatePaymentAsync(CreatePaymentRequestDto request)
    {
        var handler = handlers.FirstOrDefault(x => x.SupportedMethod == request.Method);
        
        if (handler == null)
        {
            throw new SdkException("Unsupported payment method.", HttpStatusCode.BadRequest);
        }

        CreatePaymentResponseDto response = await handler.HandleAsync(request);

        return response;
    }

    public async Task<AdditionalPaymentActionResponseDto> CancelPaymentAsync(
        AdditionalPaymentActionRequestDto requestDto)
    {
        AdditionalPaymentActionResponseDto responseSdk =
            await paymentClient.CancelPaymentAsync(requestDto, requestDto.Id);

        return responseSdk;
    }

    public async Task<AdditionalPaymentActionResponseDto> CapturePaymentAsync(
        AdditionalPaymentActionRequestDto requestDto)
    {
        AdditionalPaymentActionResponseDto responseDto =
            await paymentClient.CapturePaymentAsync(requestDto, requestDto.Id);

        return responseDto;
    }

    public async Task<AdditionalPaymentActionResponseDto> RefundPaymentAsync(
        AdditionalPaymentActionRequestDto requestDto)
    {
        AdditionalPaymentActionResponseDto responseDto =
            await paymentClient.RefundPaymentAsync(requestDto, requestDto.Id);

        return responseDto;
    }

    public async Task<PaymentDetailsResponseDto?> GetPaymentDetailsById(string id)
    {
        PaymentDetailsResponseDto? result = await paymentClient.GetPaymentDetailsById(id);

        return result;
    }
}

File: server/ruby/app/business/services/payment/payment_service.rb
module Business
  module Services
    module Payment
      class PaymentService
        attr_reader :payment_client, :handlers

        def initialize(payment_client:, handlers:)
          @payment_client = payment_client
          @handlers = handlers
        end

        def create_payment(request_dto)
          handler = handlers.find do |h|
            h.supported_method.to_s == request_dto.method.to_s
          end

          handler.handle(request_dto)
        end

        def capture_payment(request_dto)
          @payment_client.capture_payment(request_dto)
        end

        def refund_payment(request_dto)
          @payment_client.refund_payment(request_dto)
        end

        def cancel_payment(request_dto)
          @payment_client.cancel_payment(request_dto)
        end

        def get_payment_details(id)
          @payment_client.get_payment_details(id)
        end
      end
    end
  end
end
